### ISSUE TITLE
Invisible contacts letter avatar

### ISSUE THREAD

**Issue by IwishIcanFLighT** [owncloud/contacts#300]
### Expected behaviour

Contact list with different colors.
### Actual behaviour

Some contacts avatars are invisible. Selecting the list make the letter appears.

![capture](https://cloud.githubusercontent.com/assets/4090502/14061578/de212ea0-f383-11e5-861e-1eb3d04d37fa.JPG)

Also, those contact edit page looks like this:
![2](https://cloud.githubusercontent.com/assets/4090502/14061607/85fbc23e-f384-11e5-8c26-dd9cd8081ed3.JPG)
### Server configuration

**Operating system**: Debian

**Web server:** Apache

**Database:** MySQL

**PHP version:** 5.6

**ownCloud version:** 9.0.0

**Contacts version:** 1.1.0.0

**Updated from an older ownCloud or fresh install:** From 8.2.3

**Signing status (ownCloud 9.0 and above):** Got extra files errors after upgrade, deleted each files afterward.

```
No errors have been found.
```

**List of activated apps:**

```
Enabled:
  - activity: 2.2.1
  - calendar: 1.0
  - comments: 0.2
  - contacts: true
  - dav: 0.1.5
  - documents: 0.12.0
  - federatedfilesharing: 0.1.0
  - federation: 0.0.4
  - files: 1.4.4
  - files_locking: true
  - files_pdfviewer: 0.8
  - files_sharing: 0.9.1
  - files_texteditor: 2.1
  - files_trashbin: 0.8.0
  - files_versions: 1.2.0
  - files_videoplayer: 0.9.8
  - firstrunwizard: 1.1
  - gallery: 14.5.0
  - notifications: 0.2.3
  - provisioning_api: 0.4.1
  - systemtags: 0.2
  - templateeditor: 0.1
  - updatenotification: 0.1.0
Disabled:
  - encryption
  - external
  - files_external
  - files_videoviewer
  - galleryplus
  - pdflintview
  - updater
  - user_external
  - user_ldap
  - user_webdavauth
```

**The content of config/config.php:**

```
{
    "system": {
        "instanceid": "oc26ab9bf921",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "domain.com"
        ],
        "datadirectory": "\/var\/www\/domain.com\/data",
        "dbtype": "mysql",
        "version": "9.0.0.19",
        "dbname": "owncloud_db",
        "dbhost": "localhost",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "theme": "",
        "maintenance": false,
        "secret": "***REMOVED SENSITIVE VALUE***",
        "loglevel": 2,
        "forcessl": true,
__ MAIL INFO REMOVED__,
        "forceSSLforSubdomains": true,
        "trashbin_retention_obligation": "auto"
    }
}

```

**Are you using external storage, if yes which one:** No

**Are you using encryption:** No

**Are you using an external user-backend, if yes which one:** Nope

**Comment by macmeck** [comment]
Especially the contact edit form thing is pretty annoying since you cannot even read the name of the contact.

**Comment by skjnldsv** [comment]
Hum, this is strange.
Does the contacts have pictures?

Could you give me some contact uid? The color is generated from the contact uid.

**Comment by bohni** [comment]
I investigated this yesterday and just run over this issue as I wanted to create one.

The reason is that the backgorund color is generated from the contacts UID. The UID is expected to be a hash value. See `contactColor_filter.js`.

But some contacts have another UID, eg. `20131217T100925.ac9f58a2cd@example.com`

So the background-color could not be generated as the JavaScript-Method returns NaN

There is a bug report at [https://forums.bitfire.at/topic/786/vcards-of-owncloud-which-uid-s-matches-uid-are-not-synched-to-davdroid] with syncing such contacts to owncloud.

I have no idea where these UIDs come from... Is it Android? Thunderbird? Owncloud < v9.0?

Edit: Same for the header of a single shown contact.

Edit2: Is it Google? I exported my Google addressbook and imported it in my owncloud 6 server a long time ago...

**Comment by skjnldsv** [comment]
That's what I though! It's strange that you have those kinds of uids.
@DeepDiver1975 Should we override the uids in the contact app to preserve the pattern we're using?
I could also correct the filter.

**Comment by gvde** [comment]
According to RFC 2426 3.0 cards UID can have any text as value. Only with 4.0 cards you have URIs by default, but then a UID can look like this: UID:urn:uuid:f81d4fae-7dec-11d0-a765-00a0c91e6bf6...

And I think you should not modify properties of cards unless absolutely necessary. You'll never know where else this UID is used...

**Comment by skjnldsv** [comment]
@gvde this should fix it. Replace

```
var hash = input.split('-').join('');
```

by

```
var hash = input.toLowerCase().replace(/[^0-9a-f]+/g, '');
```

in /js/public/script.js (arround line 1620)

**Comment by bohni** [comment]
@skjnldsv Thanks, this fix works for me.

**Comment by Henni** [comment]
@skjnldsv we should never override any property in a vcard without the interaction of a user. This could lead to strange bugs if multiple clients are connected and try to convert the uids to their format, for example.

**Comment by skjnldsv** [comment]
That's what i though too, but wanted to ask anyway! :)

**Comment by IwishIcanFLighT** [comment]
Woah, you guys are fast!
I can't wait to test the fix! :smile: 
Thanks!

**Comment by ghost** [comment]
Just updatet to 9.0.1 (PHP 7.0.4), error still exist. Should it be solved in 9.0.1?

**Comment by skjnldsv** [comment]
No it should not.
It will be fixed in the 1.2 version of contacts.

**Comment by ghost** [comment]
perfekt, thx.
